util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+
+#define PP_STR_LEN 32
+#define PP_MAX_CTRL_MSG 64
+#define PP_CTRL_BUF_LEN 64
+#define PP_MR_KEY 0xC0DE
+
+#define INTEG_SEED 7
+#define PP_ENABLE_ALL (~0)
+#define PP_DEFAULT_SIZE (1 << 0)
+
+#define PP_MSG_CHECK_PORT_OK "port ok"
+#define PP_MSG_LEN_PORT 5
+#define PP_MSG_CHECK_CNT_OK "cnt ok"
+#define PP_MSG_LEN_CNT 10
+#define PP_MSG_SYNC_Q "q"
+#define PP_MSG_SYNC_A "a"
+
+#define PP_PRINTERR(call, retv)						\
+	do {								\
+		fprintf(stderr, "%s(): %s:%-4d, ret=%d (%s)\n", call,	\
+			__FILE__, __LINE__, (int) retv,			\
+			fi_strerror((int) -retv));			\
+	} while (0)
+
+#define PP_ERR(fmt, ...)						\
+	do {								\
+		fprintf(stderr, "[%s] %s:%-4d: " fmt "\n", "error",	\
+			__FILE__, __LINE__, ##__VA_ARGS__);		\
+	} while (0)							\
+
+int pp_debug = 0;
+
+#define PP_DEBUG(fmt, ...)						\
+	if (pp_debug) {							\
+		fprintf(stderr, "[%s] %s:%-4d: " fmt, "debug",		\
+			__FILE__, __LINE__, ##__VA_ARGS__);		\
+	}								\
+
+#define PP_CLOSE_FID(fd)						\
+	do {								\
+		int ret;						\
+		if ((fd)) {						\
+			ret = fi_close(&(fd)->fid);			\
+			if (ret)					\
+				PP_ERR("fi_close (%d) fid %d",		\
+					ret, (int) (fd)->fid.fclass);	\
+			fd = NULL;					\
+		}							\
+	} while (0)
+
+#define MAX(a,b) (((a)>(b))?(a):(b))
+
+struct ct_pingpong {
+	struct fi_info *fi_pep, *fi, *hints;
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_pep *pep;
+	struct fid_ep *ep;
+	struct fid_cq *txcq, *rxcq;
+	struct fid_mr *mr;
+	struct fid_av *av;
+	struct fid_eq *eq;
+
+	struct fid_mr no_mr;
+	struct fi_context tx_ctx, rx_ctx;
+	struct fi_context *ctx_arr;
+	uint64_t remote_cq_data;
+
+	uint64_t tx_seq, rx_seq, tx_cq_cntr, rx_cq_cntr;
+	int pp_parent_proc;
+	pid_t pp_child_pid;
+
+	fi_addr_t remote_fi_addr;
+	void *buf, *tx_buf, *rx_buf;
+	size_t buf_size, tx_size, rx_size;
+	int rx_fd, tx_fd;
+	int data_default_port;
+	char data_port[8];
+
+	char test_name[50];
+	int timeout;
+	struct timespec start, end;
+
+	struct fi_av_attr av_attr;
+	struct fi_eq_attr eq_attr;
+	struct fi_cq_attr cq_attr;
+	struct pp_opts opts;
+
+	long cnt_ack_msg;
+
+	int ctrl_port;
+	int ctrl_listenfd;
+	int ctrl_connfd;
+	struct sockaddr_in ctrl_addr;
+	char ctrl_buf[PP_CTRL_BUF_LEN + 1];
+};
+
+static const char integ_alphabet[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+static const int integ_alphabet_length = (sizeof(integ_alphabet)/sizeof(*integ_alphabet)) - 1; // without '\0'
+
+/*******************************************************************************************/
+/*                                  Compatibility methods                                  */
+/*******************************************************************************************/
+
+#ifdef __APPLE__
+int clock_gettime(clockid_t clk_id, struct timespec *tp) {
+	int retval;
+	struct timeval tv;
+
+	retval = gettimeofday(&tv, NULL);
+
+	tp->tv_sec = tv.tv_sec;
+	tp->tv_nsec = tv.tv_usec * 1000;
+
+	return retval;
+}
+#endif
+
```

Comment:
We actually thought about that, but ...
In the initial discussions, the idea was to have a self-contained source file for this new tool. I think that if we start to create an extern header/source file, other functions may be added to it, and we would come back in a way to the situation of fabtests.